### PR TITLE
bpo-32684: Fix gather to propagate cancel of itself with return_exceptions

### PR DESCRIPTION
**Pull request by 1st1**
https://bugs.python.org/issue32684

### CODE REVIEW

File: Lib/asyncio/tasks.py [hunk 1]
```diff
@@ -599,6 +600,11 @@ def cancel(self):
         for child in self._children:
             if child.cancel():
                 ret = True
+        if ret:
```
Comment (asvetlov) [inline Lib/asyncio/tasks.py:603]: I don't follow why future cancellation should be propagated but if children are tasks nothing should be done.

Comment (1st1) [reply Lib/asyncio/tasks.py:603]: See https://bugs.python.org/issue26923 and, specifically, `Task.cancel()`: if the task is done/cancelled, we don't raise `CancelledError`. Similarly for `gather()`, if all of its tasks are already done, we don't raise it.

This seems to be a super rare edge case.

Comment (asvetlov) [reply Lib/asyncio/tasks.py:603]: Got it.

**Review by asvetlov — APPROVED** [review]
lgtm

**Comment by 1st1** [comment]
3.6 PR: https://github.com/python/cpython/pull/7224

**Pull request merged**